### DISCUSSION TITLE
How do arithmetic instructions work in the code?

### DISCUSSION THREAD

**Discussion by williamistGitHub** [ArgosOfIthica/scalu#34] [Q&A] [answered]
Hello! I'm trying to understand how arithmetic instructions work in the code. It's pretty confusing and vague, both add and subtract seem to mess with bits randomly. It'd be nice if you could explain it to me as I'd rather not figure this all out myself if I don't have to. Thanks!

**Comment by ArgosOfIthaca** [comment, marked as answer]
Hello!

I agree, the backend is quite abstract, arguably a little more than it needs to be, and I'm working to simplify it. That said, addition will work the same way regardless; we implement addition the same way real hardware does, with a series of [full adders](https://fpgabeginner.com/wp-content/uploads/2020/09/threeInputTwooutputtruthtable-1024x402.png), or more specifically, the truth table associated with a full adder. The code [here](https://github.com/ArgosOfIthica/scalu/blob/master/scalu/src/backend/instructions/instruction.py#L480) is just implementing this truth table while inheriting all the mechanical overhead needed to generate the code.

**Reply by williamistGitHub** [reply]
Ah I see now, thanks.

**Reply by snoopchallenge** [reply]
you are a magician thank you
